### PR TITLE
Add RPA - Reusable Payment Addresses

### PR DESCRIPTION
**Pull request by OPReturnCode**
This pull request is a basic implementation of [https://github.com/imaginaryusername/Reusable_specs/blob/master/reusable_addresses.md](https://github.com/imaginaryusername/Reusable_specs/blob/master/reusable_addresses.md), adding support for using reusable payment addresses inside EC.

### CODE REVIEW

**Comment by fyookball** [comment]
Regarding the fc53a0f and the treatment of the ndata parameter... The whole thing is a bit convoluted because originally (in an early implementation attempt), we had the ndata parameter being passed essentially through the whole stack and the grinding was done as part of the main guts of the transaction or wallet modules.  Because that approach had poor performance, the grinding loop was moved into rpa/paycode.py and calls the transaction class directly. But, we had decided to simply leave the existing ndata parameters everywhere, at least temporarily. 

I don't know exactly what is being fixed by removing ndata from the call inside keystore.py, but I think we either need to retain these parameters if they are being used by wallet and keystore modules, or clean up the code so that ndata isn't used in those modules and only exists as a parameter at the transaction level.  In other words, its probably fine that you removed ndata as a parameter in the call to tx.sign but why even have it as a parameter in def sign_transaction in keystore.py?

**Comment by jonas-lundqvist** [comment]
(I wrote this on TG but repeating it here where it's more accessible). 

I just had a cursory look and, although I don't know the details of the spec, the main thing that stands out is the fact that a new wallet type has been introduced.
I think a more sensible way would be to have the RPA functionality a part of the `Standard` wallet. If the concern is that `scanpubkey` and/or `spendpubkey` are used for normal cashaddresses those could be generated from another derivation path. (I can't tell if that would be a real concern or not). I have a hard time thinking users would want to flip back and forth different wallets to be able to receive money.

Generated addresses could also be extended to be in the form `bitcoincash:<cashaddr>?paycode=<rpa>` for RPA enabled wallets to provide a way for a payee to provide an address where the payers wallet would use RPA on a best-effort basis but could always fall-back to the old way. With the implementation in this PR every payee would either have to *know* that the payer is using an RPA wallet or provide two different addresses (from two different wallets!). I think this is a quite serious UX issue that would severely limit the usage of this feature. Of course a payee should be able to force a `paycode:<rpa>` address if he or she wishes. This might be something more appropriate for the specification (ping @imaginaryusername).

Another thing I wonder is if there is supposed to be any subscription methods in the Electrum protocol. As I understand it a user would need to explicitly refresh the wallet in the GUI for new transactions to show up. Is there any fundamental issues that needs to be dealt with or just something that hasn't been coded yet?

**Comment by cculianu** [comment]
To be honest -- I do agree with everything you just wrote, Jonas.  :/

**Comment by OPReturnCode** [comment]
> Regarding the [fc53a0f](https://github.com/Electron-Cash/Electron-Cash/commit/fc53a0f4d26e8e8bace7e134e2af67e71684947d) and the treatment of the ndata parameter... The whole thing is a bit convoluted because originally (in an early implementation attempt), we had the ndata parameter being passed essentially through the whole stack and the grinding was done as part of the main guts of the transaction or wallet modules. Because that approach had poor performance, the grinding loop was moved into rpa/paycode.py and calls the transaction class directly. But, we had decided to simply leave the existing ndata parameters everywhere, at least temporarily.
> 
> I don't know exactly what is being fixed by removing ndata from the call inside keystore.py, but I think we either need to retain these parameters if they are being used by wallet and keystore modules, or clean up the code so that ndata isn't used in those modules and only exists as a parameter at the transaction level. In other words, its probably fine that you removed ndata as a parameter in the call to tx.sign but why even have it as a parameter in def sign_transaction in keystore.py?
I tried to go on with the decision of not removing ndata for now. The commit doesn't remove ndata from there. Just fixes the bug in the else statement that it might call a version of sign that doesn't accept ndata. You're absolutely right on removing it altogether would be a better decision. I

**Comment by OPReturnCode** [comment]
> Another thing I wonder is if there is supposed to be any subscription methods in the Electrum protocol. As I understand it a user would need to explicitly refresh the wallet in the GUI for new transactions to show up. Is there any fundamental issues that needs to be dealt with or just something that hasn't been coded yet?

That only happens with 0conf currently.

**Review by jonas-lundqvist — COMMENTED** [review]
I had a first look at this PR to get a feeling for the feature. I have not studied the actual cryptography and I did not comment on all minor issues (debug prints, access to non-validated array members, code style etc) since I think a some of the bigger design questions should be sorted out first. I have not run the code either so I have for instance no feeling if the 16 bit `prefix_size` takes seconds, minutes or hours to grind or how the UI looks.

One thing that bugs me is that this is supposedly meant to be merged to master (i.e. production ready!) before the server parts are "done". After some digging I found an unofficial Fulcrum fork with code that really doesn't look production ready. Apparently the Fulcrum part has an (undocumented) configurable `max_reusable_history` settings but there is no way for the client to know what that might be. Also the `blockchain.reusable.subscribe` RPC seems to be missing which should be very much needed for the **user expected** 0-conf to work. As I noted in one of the comments a server could also refuse short prefixes which needs to be communicated to the client somehow.

My advice: Take a step back, figure out the Electrum RPC details and then get a server implementation up and running. After that it's time for the client implementation.

File: electroncash/rpa/addr.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+"""
+Implements a custom cashaddr-style encoding for a Reusable Payment Address (RPA)
```
[the remark this answers is not in the document]
Comment (fyookball) [reply electroncash/rpa/addr.py:24]: I recall discussing this with @cculianu a while back.  I think we specifically decided to NOT consolidate it and have it be its own section of code because it didn't conform to the cashaddr spec.  I don't remember the exact reason.  Maybe some of the cashaddr error checks were at issue, as you alluded to.  Could have been because the hash size was too big.  In any case, I'm ok with having it be its own section for now.

File: electroncash/rpa/paycode.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+
+
+def _generate_privkey_from_secret(parent_privkey, secret):
+    """parent_privkey and secret are expected to be bytes
+    This function generates a receiving address based on CKD."""
+
+    return bitcoin.CKD_priv(parent_privkey, secret, 0)[0].hex()
+
+
+def get_grind_string(wallet, prefix_size="10"):
```
[the remark this answers is not in the document]
Comment (fyookball) [reply electroncash/rpa/paycode.py:148]: I don't think there's a setting.  It's just set at "10" (16 bits) in the client code.  It can be modified at the code level, but there is no UI to change this currently and that's good.

File: electroncash/rpa/paycode.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    # Decode the paycode
+    rprefix, addr_hash = addr.decode(rpa_paycode)
+    paycode_hex = addr_hash.hex().upper()
+
+    # Parse paycode
+    paycode_field_version = paycode_hex[0:2]
+    paycode_field_prefix_size = paycode_hex[2:4]
+    paycode_field_scan_pubkey = paycode_hex[4:70]
+    paycode_field_spend_pubkey = paycode_hex[70:136]
+    paycode_field_expiry = paycode_hex[136:144]
```
[the remark this answers is not in the document]
Comment (fyookball) [reply electroncash/rpa/paycode.py:203]: @OPReturnCode we could consider implementing the validation of this when the user tries to send, if its not too much work.

File: electroncash/rpa/paycode.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+
+        grind_nonce_string = str(grind_count)
+        grinding_message = paycode_hex + grind_nonce_string + grinding_version
+        ndata = sha256(grinding_message)
+        # Re-sign the transaction input.
+        tx._sign_txin(0, 0, sec, compressed, use_cache=False, ndata=ndata)
+
+        if progress_callback and progress_count < grind_count // 1000:
+            progress_count = grind_count // 1000
+            progress_callback(progress_count)
```
[the remark this answers is not in the document]
Comment (fyookball) [reply electroncash/rpa/paycode.py:320]: I don't know about "clever" but I did do some trial-and-error estimation in terms of choosing values that cause the grinding to finish before the progress bar hits 100%....most of the time.

File: electroncash/rpa_manager.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+        # Not part of the normal peristent loop.  This is called externally when the wallet 
+        # wants to check the mempool.  We make the request similar to the normal phase 1 
+        # and let the module do the rest.
+                
+        # Define the "grind string" (the RPA prefix)
+        rpa_grind_string = self.wallet.get_grind_string()
+        params = [rpa_grind_string]
+        requests = []
+        requests.append(('blockchain.reusable.get_mempool', params))
```
[the remark this answers is not in the document]
Comment (fyookball) [reply electroncash/rpa_manager.py:67]: I don't see why this needs to be handled at all on the client side.  Let the client request whatever it wants.  Like the spec says, the server can limit excessively short prefixes.

File: electroncash/rpa_manager.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+            
+        # Define height variables.     
+        server_height = self.network.get_server_height()
+
+        if not server_height:
+            return
+
+        rpa_height = self.wallet.storage.get('rpa_height')
+        if rpa_height is None:
+            rpa_height = server_height - 100
```
[the remark this answers is not in the document]
Comment (fyookball) [reply electroncash/rpa_manager.py:92]: Yea good point we probably need the user to specify the blockheight to scan when doing a restore.  The idea was set it just below the tip for new wallets but this doesn't cover restores.

File: electroncash/rpa/addr.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cculianu) [reply electroncash/rpa/addr.py:24]: Honestly @jonas-lundqvist is right.  The copy-pasta here is just lazy and terrible.

Don't invoke my name here.  I never agreed to this.  I just didn't stop you from doing it for want of defending the rest of the codebase from modification (and potential breakage).

**Comment by imaginaryusername** [comment]
> I had a first look at this PR to get a feeling for the feature. I have not studied the actual cryptography and I did not comment on all minor issues (debug prints, access to non-validated array members, code style etc) since I think a some of the bigger design questions should be sorted out first. I have not run the code either so I have for instance no feeling if the 16 bit `prefix_size` takes seconds, minutes or hours to grind or how the UI looks.
> 
> One thing that bugs me is that this is supposedly meant to be merged to master (i.e. production ready!) before the server parts are "done". After some digging I found an unofficial Fulcrum fork with code that really doesn't look production ready. Apparently the Fulcrum part has an (undocumented) configurable `max_reusable_history` settings but there is no way for the client to know what that might be. Also the `blockchain.reusable.subscribe` RPC seems to be missing which should be very much needed for the **user expected** 0-conf to work. As I noted in one of the comments a server could also refuse short prefixes which needs to be communicated to the client somehow.
> 
> My advice: Take a step back, figure out the Electrum RPC details and then get a server implementation up and running. After that it's time for the client implementation.

I didn't pay much attention to this PR because I was busy somewhere else, but from a high level if a server isn't ready that means the feature isn't functional, no matter the state of the rest of the PR. It'll be terrible if this is merged, advertised as RPA being working in some shape, then having eager users angry at false advertisement the moment they dive a bit in. 

It's fine to have this sitting around as an attempt that can be used as basis to construct a better version later, but definitely do not merge any version of this until a server can be run from somewhere.

**Comment by imaginaryusername** [comment]
> 
> Generated addresses could also be extended to be in the form `bitcoincash:<cashaddr>?paycode=<rpa>` for RPA enabled wallets to provide a way for a payee to provide an address where the payers wallet would use RPA on a best-effort basis but could always fall-back to the old way. With the implementation in this PR every payee would either have to _know_ that the payer is using an RPA wallet or provide two different addresses (from two different wallets!). I think this is a quite serious UX issue that would severely limit the usage of this feature. Of course a payee should be able to force a `paycode:<rpa>` address if he or she wishes. This might be something more appropriate for the specification (ping @imaginaryusername).

I agree that an extension of normal cashaddr to provide for an optional RPA extension would be a smoother UX, but the chief purpose of having RPA as a payee is generally to conceal your payment history, whether in public or through another layer of identity service. I predict the latter would ultimately be the most prevalent usecase, making this discussion obsolete; but even for the former (post address in public) usecase, having it "optional" will defeat its purpose as a large part of your history is revealed anyway. 

Basically: If someone can deal with posting a `bitcoincash:xxxx?paycode=yyy` address, he should just post a `bitcoincash:xxxx` address anyway and skip the RPA hassle. I struggle to think of a case where this optional extension is useful.

**Comment by cculianu** [comment]
> I didn't pay much attention to this PR because I was busy somewhere else, but from a high level if a server isn't ready that means the feature isn't functional, no matter the state of the rest of the PR. It'll be terrible if this is merged, advertised as RPA being working in some shape, then having eager users angry at false advertisement the moment they dive a bit in.
> 
> It's fine to have this sitting around as an attempt that can be used as basis to construct a better version later, but definitely do not merge any version of this until a server can be run from somewhere.

THANK YOU!! @imaginaryusername .. I knew I loved ❤️  you. I forgot why. It's moments like these when you can be impartial and extremely rational!!

As an olive branch to all those involve: I offer to also spend time on this PR and see if I can add some value to it as well. But THANK YOU for chiming in here! I absolutely love you!

❤️ ❤️ ❤️ ❤️ ❤️ ❤️

**Comment by imaginaryusername** [comment]
...does a server now exist making this actually usable?!

If not, why is this merged? I pull from master for my daily use wallet, should I be concerned?

**Comment by fyookball** [comment]
The fountainhead server should be operational.  I'm checking on that.  

I don't think this will impact non-RPA wallet types whatsoever, so no need for concern.
